tests/verify.rs

```diff
@@ -1,8 +1,170 @@
 use std::convert::TryFrom;
 
-use webpki::{EndEntityCert, KeyUsage, SubjectNameRef, Time};
+use rcgen::{
+    BasicConstraints, Certificate, CertificateParams, DnType, GeneralSubtree, IsCa,
+    KeyUsagePurpose, NameConstraints,
+};
+use webpki::{EndEntityCert, Error, KeyUsage, SubjectNameRef, Time};
+use x509_parser::extensions::{GeneralName, NameConstraints as X509ParserNameConstraints};
+use x509_parser::prelude::FromDer;
+
 use webpki_roots::TLS_SERVER_ROOTS;
 
+#[test]
+fn name_constraints() {
+    for name_constraints in TLS_SERVER_ROOTS.iter().filter_map(|ta| ta.name_constraints) {
+        let time = Time::from_seconds_since_unix_epoch(0x40000000); // Time matching rcgen default.
+        let test_case = name_constraints_testcase(name_constraints);
+        let trust_anchors =
+            &[webpki::TrustAnchor::try_from_cert_der(&test_case.trust_anchor).unwrap()];
+
+        // Each permitted EE should verify without error.
+        for permitted_ee in test_case.permitted_certs {
+            webpki::EndEntityCert::try_from(permitted_ee.as_slice())
+                .unwrap()
+                .verify_for_usage(
+                    ALL_ALGORITHMS,
+                    trust_anchors,
+                    &[],
+                    time,
+                    KeyUsage::server_auth(),
+                    &[],
+                )
+                .unwrap();
+        }
+
+        // Each forbidden EE should fail to verify with the expected name constraint error.
+        for forbidden_ee in test_case.forbidden_certs {
+            let result = webpki::EndEntityCert::try_from(forbidden_ee.as_slice())
+                .unwrap()
+                .verify_for_usage(
+                    ALL_ALGORITHMS,
+                    trust_anchors,
+                    &[],
+                    time,
+                    KeyUsage::server_auth(),
+                    &[],
+                );
+            assert!(matches!(result, Err(Error::NameConstraintViolation)));
+        }
+    }
+}
+
+struct ConstraintTest {
+    trust_anchor: Vec<u8>,
+    permitted_certs: Vec<Vec<u8>>,
+    forbidden_certs: Vec<Vec<u8>>,
+}
+
+fn name_constraints_testcase(webpki_name_constraints: &[u8]) -> ConstraintTest {
+    // Create a trust anchor CA certificate that has the name constraints we want to test.
+    let mut trust_anchor = CertificateParams::new([]);
+    trust_anchor
+        .distinguished_name
+        .push(DnType::CommonName, "Name Constraint Test CA");
+    trust_anchor.is_ca = IsCa::Ca(BasicConstraints::Unconstrained);
+    trust_anchor.key_usages = vec![
+        KeyUsagePurpose::KeyCertSign,
+        KeyUsagePurpose::DigitalSignature,
+    ];
+    let name_constraints = rcgen_name_constraints(webpki_name_constraints);
+    trust_anchor.name_constraints = Some(name_constraints.clone());
+    let trust_anchor = Certificate::from_params(trust_anchor).unwrap();
+
+    fn certs_for_subtrees<'a>(
```

Comment:
Maybe this could become a closure instead, retaining access to the outer `trust_anchor` and `name_constraints.permitted_subtrees`?

The `name_fn` argument could be respun as a `suffix: &str` argument instead, perhaps?

Reply:
Good idea, that's much tidier. Updated :rocket: